src/core/stream/representation/check_for_discontinuity.ts

```diff
@@ -78,105 +81,141 @@ export default function checkForDiscontinuity(
+
+    // and no segment will fill in that hole
+    (nextSegmentStart === null ||
+     nextBufferedSegment.infos.segment.end <= nextSegmentStart)
+  ) {
+    log.error("RS: current discontinuity encountered",
+              adaptation.type, nextBufferedSegment.bufferedStart);
+    return { start: undefined,
+             end: nextBufferedSegment.bufferedStart };
+  }
+
+  // Check if there's a discontinuity BETWEEN segments of the current range
+  let nextHoleIdx;
+  for (
+    let bufIdx = nextBufferedInRangeIdx + 1;
+    bufIdx < bufferedSegments.length;
+    bufIdx++)
+  {
+    const currSegment = bufferedSegments[bufIdx];
+    const prevSegment = bufferedSegments[bufIdx - 1];
+
+    // Exit as soon we miss information or when we go further than `checkedRange`
+    if (currSegment.bufferedStart === undefined ||
+        prevSegment.bufferedEnd === undefined ||
+        currSegment.bufferedStart >= checkedRange.end)
     {
-      log.error("RS: current discontinuity encountered",
-                adaptation.type, nextBufferedSegment.bufferedStart);
-      return { start: undefined,
-               end: nextBufferedSegment.bufferedStart };
-    } else {
-      let nextHoleIdx;
-      for (
-        let bufIdx = nextBufferedInRangeIdx + 1;
-        bufIdx < bufferedSegments.length;
-        bufIdx++)
-      {
-        const currSegment = bufferedSegments[bufIdx];
-        const prevSegment = bufferedSegments[bufIdx - 1];
-        if (currSegment.bufferedStart === undefined ||
-            prevSegment.bufferedEnd === undefined ||
-            currSegment.bufferedStart >= checkedRange.end)
-        {
+      break;
+    }
+
+    // If there is a hole between two consecutive buffered segment
+    if (currSegment.bufferedStart - prevSegment.bufferedEnd > 0) {
+      nextHoleIdx = bufIdx;
+      break;
+    }
+  }
+
+  // If there was a hole between two consecutives segments, and ifh this hole
+  // comes before the next segment to load, there is a discontinuity (that hole!)
+  if (nextHoleIdx !== undefined &&
+      (nextSegmentStart === null ||
+       bufferedSegments[nextHoleIdx].infos.segment.end <= nextSegmentStart))
+  {
+    const start = bufferedSegments[nextHoleIdx - 1].bufferedEnd as number;
+    const end = bufferedSegments[nextHoleIdx].bufferedStart as number;
+    log.error("RS: future discontinuity encountered", adaptation.type, start, end);
+    return { start, end };
+
+  } else if (nextSegmentStart === null) {
+    // If no hole between segments and no segment to load, check for a
+    // discontinuity at the end of the Period
+
+    if (hasFinishedLoading && period.end !== undefined) { // Period is finished
+      if (checkedRange.end < period.end) { // We've not reached the Period's end yet
+        return null;
+      }
+
+      // Check if the last buffered segment ends before this Period's end
+      // In which case there is a discontinuity between those
+      let lastBufferedInPeriodIdx;
```

Comment:
I suggest that it may clearer to add function for each time you assign values to this type of variables (lastBufferedInPeriodIdx, nextHoleIdx, nextBufferedInRangeIdx).
I mean :
const lastBufferedInPeriodIdx = getLastBufferedInPeriodIdx(args);

Maybe the checkForDiscontinuity may be clearer as it.

Reply:
Done